RFD 98 - Secure OCI images (#19045)
gravitational/teleport: rfd/0096-securing-docker-images.md

> **NOTE:** We are _already_  using Distroless images to distribute some Teleport 
> plugins. This would extend their use to Teleport proper.

### 2. Ongoing scanning

Using an _ongoing_ automated scanner means that we do not just check for 
vulnerabilities at image creation time, instead we proactively & _continually_ scan
for any vulnerabilities  that may be discovered until the image is either replaced 
with a newer verion of the image, or the support lifetime for the verion of Teleport 
on that image expires (i.e. falls out of our 3-version support window)

## Implementation Details

## 1. Image construction

### Teleport Image Requirements

What is the minimal set of requirements to run Teleport on Linux in a container?

Most Teleport dependencies are statically compiled into the `teleport` binary, giving us a 
smaller set of runtime dependencies than you might imagine:

   1. Teleport
   2. `GLIBC` >= 2.17
   3. `dumb-init` is required for correct signal and child processes handling
      inside a container.
   4. `libpam` (and its transitve dependencies) for PAM support 
   5. CA certificates

> 

Author: I might just change it to _C runtime libraries_ to cover both

      in the build process when `distroless` is used as an external dependency in 
      an enclosing `bazel` workspace, requiring manual intervention in the build to
      solve.

   Using `bazel` does not resolve a major limitation of using a basic 
   `Dockerfile` (i.e. the `xcopy` style install) and introduces more complexity, in 
   terms of both build process and tooling, so was rejected in favour of the 
   `Dockerfile` approach.

 * **[apko](https://github.com/chainguard-dev/apko)**: Apko is a tool for quickly 
   building minimalist, reproducible Alpine linux images, using a declarative 
   format. While I found the tool verly neat, the images it generates are still closer 
   to a "debug" Distroless image. 

   Using `apko` would also require us to build an Alpine linux package for Teleport 
   to integrate it nto the build.

   I seriously considered recommending `apko`, as it has some neat features (e.g.
   automatically producting a SBOM as part of the construction process), but in the 
   end I rejected it because of the extra software included in the resulting images.

> 

Reviewer: We _could_ consider rolling our own base image (with a normal Dockerfile, without bazel). This would have some major downsides (primarily the effort required to both build the tooling to create the base image and maintain it), but it could also have some non-technical business advantages:
* If we can solve some (or all) of the issues with other tools above then providing a secure, simple image to the open source community could help expand our brand.
* We could potentially integrate security fixes for upstream packages faster, showing our commitment to security to potential customers.
* We could implement features like SBOM reports into our toolchain which would add some transparency for both Gravitational and any customers using our Teleport image.
* By implementing the upstream image ourselves we can insure that all current and future products can easily adhere to our security standards, without reinventing the wheel for each one. This is doubly so when taking multi-arch images into account, which introduces a lot of additional technical problems for each image to solve.

I don't know that this route conforms to our high level strategic goals, but I thought it would be worth considering.

    fail_build[Fail the build]
    sign_image[$ cosign ...]
    promote_build[/Release engineer<br/>promotes build/]
    promote_image[Candidate image<br/>becomes release image<br/>in Public ECR]
    build_ok?{Does all go well<br/>with the rest of<br/>the build?}
    
    pull_base --> signature_valid?
    signature_valid? -- yes --> base
    signature_valid? -- no --> fail_build
    deb --> build
    df --> build
    base --> build
    build --> sign_image --> candidate_image
    candidate_image --> push_internal
    push_internal --> trivy_scan
    trivy_scan --> upload_results 
    upload_results --> bad_scan?
    bad_scan? -- no --> build_ok?
    bad_scan? -- yes --> fail_build
    promote_build --> promote_image
    build_ok? -- yes --> promote_build
    build_ok? -- no --> fail_build
```    

### Debug Images

Troubleshooting a distroless image is hard, as there are no tools baked into
the image to aid in debugging a deployment.

In addtion to the main distroless image sets, the `distroless` team also supplies 
a `debug`-tagged image that includes `busybox` (that is, a basic shell and some
utilities). In order to provide tooling for troubleshooting a Teleport installation,

> we will create a parallel `teleport-debug` image, based on a distroless `debug` image.

Reviewer: Have we considered going the opposite route, where we add a `minimal` tag that is designed to have only a few packages, updated regularly? This may result in less breaking changes for our customers.

While we should take as much care as possible when constructing and 
monitoring this image, use of the debug image should probably be
considered "at your own risk".

### Compatibility Guarantees

We have clients relying on the existing behaviour (and contents) of our images. We
should treat releasing these distroless images as a compatibility break, and make 
our customers aware of our intentions well in advance so that they can prepare.

## 2. Scanning and monitoring the image

### Collecting the data 

> We are already using the scanning tools built-ino AWS ECR, which periodically scans our 

Reviewer: ```suggestion
We are already using the scanning tools built-in to AWS ECR, which periodically scans our 
```

images and results of the scan into our Panther SIEM instance.

To increase the size and quality of the vulnetrability database we scan with, we should 
also use `trivy` for ongoing scans, in addition to the build-time scans described above. 

Trivy's in a GHA integration should make it straightforward to run 
workflow to repeatedly scan out published images.

The output of the scan will be injected into the GitHub code scanning tab, and our 
Panther instance can then ingest the alerts via the GitHub audit log.

### Reacting to alerts

> Once the allert data is aggregated into Panther, we can configure events to lodge 

Reviewer: ```suggestion
Once the alert data is aggregated into Panther, we can configure events to lodge 
```

GitHub issues and/or alert the development team via Slack or e-mail. The development 

> team will be expected to resolve the issues as per the eleport Vulnerability Management 

Reviewer: ```suggestion
team will be expected to resolve the issues as per the Teleport Vulnerability Management 
```

---
authors: Trent Clarke (trent@goteleport.com)
state: draft
---

# RFD 0096 - Delivering secure Teleport OCI Images

# Required Approvers
* Engineering: @r0mant

> * Security: @reedloden

Reviewer 2: ```suggestion
* Security: (@reedloden && @jentfoo)
```

   to integrate it nto the build.

   I seriously considered recommending `apko`, as it has some neat features (e.g.
   automatically producting a SBOM as part of the construction process), but in the 
   end I rejected it because of the extra software included in the resulting images.

### Smoke testing

A smoke test is a simple, quick test to assert basic functionality. We simply want to 
find out if Teleport will even start in the environment contained by our container 
image. While this is not _per se_ a security matter, using a distroless base image 
means we have very little padding if an unexpected dependency finds its way into
Teleport. 

We want to stop ourselves shipping garbage to our customers if at all possible.

### Image signing

In order to allow our customers to validate our published Teleport images, 
our images will be signed using the `cosign` tool, similarly to the Distroless 
base images.

The `cosign` tool [integrates well with GHA](https://github.blog/2021-12-06-safeguard-container-signing-capability-actions/), 
and is even included in the template "how to publish a docker image" example 

> workflow. It also provides "keyless" signing via OIDC, meaning that our build

Reviewer 3: @tcsc As part of our automatic upgrades work we want our Kube Operator to be able to pull down an image and verify its signature. According to the documentation, it seems like keyless signing requires the use of OIDC?

https://docs.sigstore.dev/cosign/sign/#keyless-signing

> You can use Cosign to sign without keys by authenticating with an OIDC (OpenID Connect) protocol supported by Sigstore. Currently, you can authenticate with Google, GitHub, or Microsoft.
> This will open a browser window to authenticate your credentials for the signature.

In which case the Operator wouldn't be able to do that? Unless I'm misunderstanding how it works.

Should we use signing with regular keys instead and host the public key somewhere Operator can fetch it and verify the signature?

cc @fspmarshall @hugoShaka @reedloden

Reviewer 3: Looks like keyless signing verification doesn't require OIDC, just access to the public transparency log so operator should still be able to verify it.

Would be nice to try it out though and confirm - @tcsc do you have any signed images published anywhere so @hugoShaka can use them as examples as he's working on building the upgrade operator?

Reviewer 4: The Kubernetes updater will be deployed once and should be able to operate for 1+ year without being updated. As Keyless signing is experimental and can know breaking changes in the near future, I think it is a bit risky to rely only on Keyless signing for such a long-lived piece of software.

Keyless is way less costly than setting up a secure signing system, I suggest we use it by default but still generate a failover signing key pair, store it somewhere where no one has access and bake it in the validation policies/binaries we're going to ship.

This way, if we have an issue with keyless, have to move out of GitHub actions, Fulcio is down, or face a major CI incident, we would be able to break the glass and start using the keypair instead of asking everyone to update their policy/validation mechanism.

This increases the attack surface as we would have an additional secret to secure, but I think the tradeoff could be positive if we consider the availability and resiliency guarantees we gain.

This might also address some of @wadells [concerns](https://github.com/gravitational/teleport/pull/19045/files#r1072712358).

Reviewer 5: > This might also address some of @wadells [concerns](https://github.com/gravitational/teleport/pull/19045/files#r1072712358).

Yeah -- a key we control would certainly alleviate my future-proofing concern.  My experience is that Teleport CI/Release systems have a 3-4 year lifespan, and thus I'm hesitant to have a tight coupling between the build system and the customer validation UX.  I'd much prefer a design where we can plug the build infra without any change needed on the consumer side.

    sbom[Generate SBOM with syft]
    attach_sbom[Attach SBOM to<br/>Candidate Image with cosign]
    
    pull_base --> signature_valid?
    signature_valid? -- yes --> base    
    signature_valid? -- no --> fail_build
    teleport_deb --> build
    third_party_debs --> build
    df --> build
    base --> build    
    build --> candidate_image

    candidate_image --> smoke_test   
    smoke_test --> smoke_test_pass?
    smoke_test_pass? -- yes --> push_internal
    smoke_test_pass? -- no --> fail_build
    push_internal --> trivy_scan
    push_internal --> sbom
    trivy_scan --> upload_results 
    push_internal --> sign_image
    sbom --> attach_sbom
    upload_results --> code_scanning_alerts[(GitHub Code<br/>Scanning alerts)]
    code_scanning_alerts --> panther[(Panther SIEM)]
    
    sign_image --> build_ok?
    upload_results --> build_ok?
    attach_sbom --> build_ok?
    
    promote_build --> promote_image
    build_ok? -- yes --> promote_build
    build_ok? -- no --> fail_build
```    

### Debug Images

Troubleshooting a distroless image is hard, as there are no tools baked into
the image to aid in debugging a deployment.

> In addtion to the main distroless image sets, the `distroless` team also supplies 

Reviewer 3: ```suggestion
In addition to the main distroless image sets, the `distroless` team also supplies 
```

> a `debug`-tagged image that includes `busybox` (that is, a basic shell and some

Reviewer 6: If we hold the debug images to the same standard as defined above, we will lose some of the benefits of reduced internal efforts (more surface area potentially requiring prompt responses).

I agree distroless is hard to debug and we should consider this.  It may be worth reconsidering `apko` or other options that provide a single, minimized, but debugable distro rather than trying to hold two release streams to the same rigid standard.

Reviewer 7: Right. This is a bit of a worry that we would say that our happy path image is secure, but unhappy path is "use at your own risk". Attacker would always choose the path of least resistance. They would attempt to infiltrate the weak link - the debug container, and patiently wait for the Teleport's crash on someone infrastructure.

That is why I think we should provide a debug container that is as secure as our production. Can we use apko to generate a secure debug container? If that's the case, may be reconsider apko for the main use-case as well?

TLDR: Attackers will pick our weakest link, which is a debug container.

Reviewer 3: I may be missing something but our debug image will also be based on distroless and signed, will just include extra tools like busybox. Apko also generates images with extra stuff in them but it's a some lesser-known 3rd party tool while our current debug image will be based on Google's official distroless debug image which sounds like a better choice? I guess the downside is that we have 2 images but I don't think it's fair to say that the "debug" image (maybe name "debug" is a bit unfortunate) is held to a lesser standard than the "slim" image. @tcsc correct me if I'm wrong.

Reviewer 6: To better articulate the advantages I see in going distroless:
1. Smaller in size (less storage and transfer costs)
2. Reduced footprint results in less dependency CVE's that require team interruptions to address
3. Less tooling for attackers to be able to utilize once an escape has occurred
4. Easier to harden (basically already there)

Using a `debug` image compromises each one of these qualities (to varying degrees):
1. More unused contents that we have to store and transfer
2. More tools, which if we hold `debug` to the same standard could result in more team interruptions to keep it updated and secure
3. Just the availability of a shell is meaningful, but other GNU tooling is also present
4. The act of hardening a `debug` image may render the `debug` capability hindered to the point that it could never adhere to the same standard of security

TL:DR; The act of hardening and keeping `debug` up to date will be more effort than if we have a single image that we focus our efforts on.  I really like `distroless`, but I dislike the idea of eroding it's benefits by needing to maintain a second `debug` image.

That said, there is no technical limitation, or unmanageable security risk here.  I just believe that this solution will result in more effort for the team than if we adopted a single base image solution.

Reviewer 7: @jentfoo @r0mant if our `debug` image will also use distroless, signed and scanned, I have no objections to this plan. But we should not say `use at your own risk` then. We will be holding our debug images to the same standard, which I think is appropriate. 

If that's the case, I have no objections.

Reviewer 5: Taking a step back, our motivation for the debug image is:

[This is what we're offering to customer today](https://github.com/gravitational/teleport/blob/e4c97db936239ca3bac1d8c53956829c7ebb64c9/build.assets/charts/Dockerfile) -- a fat image with useful debugging tools in it.

We want to stop doing that by default, but we can't have an instant cutover.  Therefore, we *must* offer a image compatible with what were putting out today in the short term.  It won't be up to the same hardening standards as the minimial images -- by design.

Reviewer 6: Sounds good to me, particularly if the long term strategy is to have no debug image.  I just was trying to highlight that if we feel a debug version is necessary, we may want to consider just focusing our efforts on a single, secure, but debug-able image.

utilities). In order to provide tooling for troubleshooting a Teleport installation,
we will create a parallel `teleport-debug` image, based on a distroless `debug` image.

While we should take as much care as possible when constructing and 
monitoring this image, use of the debug image should probably be
considered "at your own risk".

### Compatibility Guarantees

> We have clients relying on the existing behaviour (and contents) of our images. We

Reviewer 3: Probably means this is going into Teleport 13 now?

should treat releasing these distroless images as a compatibility break, and make 
our customers aware of our intentions well in advance so that they can prepare.

## 2. Maintaininece 

### Collecting data 

We are already using the scanning tools built-in to AWS ECR, which periodically scans our 
images and results of the scan into our Panther SIEM instance.

> To increase the size and quality of the vulnetrability database we scan with, we should 

Reviewer 3: ```suggestion
To increase the size and quality of the vulnerability database we scan with, we should 
```

also use `trivy` for ongoing scans of our published (i.e. public) images, in addition
to the build-time scans described above. 

Trivy's in a GHA integration should make it straightforward to run 

> workflow to repeatedly scan our published images on a regular schedule.

Reviewer 3: How fast/slow is it? Can it also run on respective Dockerfile modifications?

The output of the scan will be injected into the GitHub code scanning system. From there 
the alerts can be picked up buy our Panther SIEM, which already integrates into our 
GitHub account. 

### Reacting to alerts

> Once the alert data is aggregated into Panther, we can configure events to lodge 

Reviewer 6: @reedloden @adaadb6 @wadells, Thoughts on if the security team should triage these alerts first, or should the dev team?

I prefer these alerts shift left to the dev team first, but Panther may not be the easiest tool to do that.  Thoughts?

Reviewer 8: Just to follow up on this, it is doable with Panther. We don't have a dedicated SOAR but we can send alerts base on tags or certain attributes to a specific channel.

Reviewer 5: Hmmm... lets consider our different container use cases:

* **Vuln found in a not-yet-published image**:  This should be shifted left, and ideally the dev team.  However we still need to consider the SLA and compensating factors -- so we probably can't have a hardline "nothing ships with a known vuln policy -- there needs to be an override mechanism to mark some automation flagged vulns as acceptable.
* **New vuln found in a previously published image**:
  * If the image is one of our fully versioned images (e.g. `teleport:12.0.2`), we currently do not change these artifacts after they're published, so there is nothing to be done. Public ECR doesn't allow tag immutability, but our private ECR where these are staged has tag immutablity turned on -- preventing these from ever being updated.
  * If the image is one of our floating tags (e.g. `teleport:12`), we want to know and make sure the vuln is fixed within the SLA.  I'd imagine we _want_ to treat these like any other vulnerability tracked in the security vulnerabilities project.  Having it go through Panther and to us first seems fine to me.

* **If the image is a `debug` image**: We can probably skip scanning and alerting on these.  They're expected to have a lot more packages and are intentionally classified as less secure.

In short, a bit of column A and a bit of column B

   automatically producting a SBOM as part of the construction process), but in the 
   end I rejected it because of the extra software included in the resulting images.

### Smoke testing

A smoke test is a smmple, quick test to assert basic functionality. We simply want to 
find out if Teleport will even start in the environment contained by our container 
image. While this is not _per se_ a security matter, using a distroless base image 
means we have very little padding if an unexpected dependency finds its way into
Teleport. 

> We want to stoip ourselves shipping garbage to our customers if at all possible.

Reviewer: ```suggestion
We want to stop ourselves shipping garbage to our customers if at all possible.
```

also use `trivy` for ongoing scans of our published (i.e. public) images, in addition
to the build-time scans described above. 

Trivy's in a GHA integration should make it straightforward to run 
workflow to repeatedly scan our published images on a regular schedule.

The output of the scan will be injected into the GitHub code scanning system. From there 
the alerts can be picked up buy our Panther SIEM, which already integrates into our 
GitHub account. 

### Reacting to alerts

Once the alert data is aggregated into Panther, we can configure events to lodge 
GitHub issues and/or alert the development team via Slack or e-mail. The development 
team will be expected to resolve the issues as per the Teleport Vulnerability Management 
Policy.

### Periodic rebuilding

Our current process rebuilds the docker images once a day, using the same 
sources and build artifacts used in the original published release. This is 
to ensure that any updates to the underlying base image are quickly and automatically 
integrated into the released image. 

We should continue rebuilding these images daily, but incorporate the same 

> supply chain checks as with the release builds descriobed above (e.g. verifying

Reviewer: ```suggestion
supply chain checks as with the release builds described above (e.g. verifying
```
